### ISSUE TITLE
Use Vector3d and std::tuple in force functions

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#2502]
Use `Vector3d force` instead of `double force[3]` and `std::tuple` instead of output arguments in bonded/non-bonded functions called by force/pressure functions, i.e. from

```c++
// /src/core/bonded_interactions/umbrella.hpp
inline int calc_umbrella_pair_force(
    Particle const *p1, Particle const *p2,
    Bonded_ia_parameters const *ia_params,
    double const d[3], double force[3]) {
   ...
}
```
```c++
// /src/core/forces_inline.hpp
double force[3] = {0., 0., 0.};
...
bond_broken = calc_umbrella_pair_force(p1, p2, iaparams, dx, force);
```

to

```c++
#include <tuple>
inline std::tuple<int, Vector3d> calc_umbrella_pair_force(
    Particle const *p1, Particle const *p2,
    Bonded_ia_parameters const *ia_params,
    Vector3d const d) {
   ...
}
```
```c++
#include <tuple>
Vector3d force;
...
std::tie(bond_broken, force) = calc_umbrella_pair_force(p1, p2, iaparams, dx);
```

Maybe also use `bool` instead of `int` for  variable `bond_broken` and the return value of bonded forces.

**Comment by jngrad** [comment]
C++14:
```c++
#include <tuple>

inline std::tuple<bool, Vector3d>
calc_pair_force(Particle const *p1, Particle const *p2,
                Bonded_ia_parameters const *ia_params) {
  if(d < cutoff) {
    return std::make_tuple(true, Vector3d{});
  }
  Vector3d force;
  /* ... */
  return std::make_tuple(false, force);
}

void main_loop() {
  Vector3d force;
  bool bond_broken;
  std::tie(bond_broken, force) = calc_pair_force(p1, p2, iaparams, dx);
}
```

C++17:
```c++
#include <tuple>

inline std::tuple<bool, Vector3d>
calc_pair_force(Particle const *p1, Particle const *p2,
                Bonded_ia_parameters const *ia_params) {
  if(d < cutoff) {
    return {true, {}};
  }
  Vector3d force;
  /* ... */
  return {false, force};
}

void main_loop() {
  auto [bond_broken, force] = calc_pair_force(p1, p2, iaparams, dx);
}
```

**Comment by fweik** [comment]
The same should be done for the energy kernels.

**Comment by jngrad** [comment]
Target:
```c++
boost::optional<double> tab_angle_energy(...);
boost::optional<std::tuple<Vector3d, Vector3d, Vector3d>> calc_tab_angle_force(...);
```
Progress: [fix-2502-p1b...fix-2502-p2](https://github.com/jngrad/espresso/compare/fix-2502-p1b...fix-2502-p2)

Full benchmarks LJ:

|  commit |           label |  empty | default | maxset |
| ------- | --------------- | ------ | ------- | ------ |
| ac26346 | boost::optional |  64.9% |  106.5% | 111.9% |
| eb89f37 |      std::tuple |  64.3% |  104.8% | 113.7% |
| f1a8a0c |    fix-2502-p1b |  62.3% |  100.6% | 104.1% |
| 158ce1d |    fix-2502-p1a |  62.0% |  100.8% | 105.6% |
| e93fcb3 |     python base |  61.8% |  100.0% | 105.1% |

Full benchmarks P3M:

|  commit |           label |  empty | default | maxset |
| ------- | --------------- | ------ | ------- | ------ |
| ac26346 | boost::optional |  95.6% |  105.0% | 118.0% |
| eb89f37 |      std::tuple |  99.4% |  104.8% | 106.5% |
| f1a8a0c |    fix-2502-p1b |  86.4% |  100.8% | 102.2% |
| 158ce1d |    fix-2502-p1a |  85.3% |  106.1% | 101.7% |
| e93fcb3 |     python base |  87.6% |  100.0% |  95.7% |

Confidence interval is ±0.2%

**Comment by RudolfWeeber** [comment]
Am I misreading the P3m numbers for the Python branch? Why is maxset faster than default?

**Comment by jngrad** [comment]
I'm not sure. The base commit was benchmarked on a different day, with multiple docker containers open, but none of them were running anything. I can re-run that commit plus another commit overnight (4 hrs/commit), but I've got different programs idle in the background today. At some point we should consider setting up a dedicated machine for benchmarks to avoid day-to-day variance.

**Comment by jngrad** [comment]
Maybe something was running on my machine 2 days ago, but it couldn't have been condor jobs because I always put my machine on drain beforehand. Anyway, here are the new benchmarks:

LJ:

|  commit |           label |  empty | default | maxset |
| ------- | --------------- | ------ | ------- | ------ |
| ac26346 | boost::optional |  64.6% |  103.9% | 111.1% |
| f1a8a0c |    fix-2502-p1b |  62.3% |  100.4% | 104.5% |
| e93fcb3 |          python |  61.5% |  100.0% | 104.4% |

P3M:

|  commit |           label |  empty | default | maxset |
| ------- | --------------- | ------ | ------- | ------ |
| ac26346 | boost::optional |  96.0% |  105.1% | 108.7% |
| f1a8a0c |    fix-2502-p1b |  91.2% |  102.5% | 106.8% |
| e93fcb3 |          python |  83.9% |  100.0% | 101.1% |

**Comment by fweik** [comment]
We already decided to go with the optional. Non of these benchmarks suggest that there are drastic performance penalties, so nothing changes.
